eskipfile/remote.go

```diff
@@ -0,0 +1,146 @@
+package eskipfile
+
+import (
+	"errors"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/routing"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type remoteEskipFile struct {
+	preloaded       bool
+	remotePath      string
+	localPath       string
+	eskipFileClient *WatchClient
+	threshold       int
+	verbose         bool
+}
+
+// RemoteWatch creates a route configuration client with (remote) file watching. Watch doesn't follow file system nodes,
+// it always reads (or re-downloads) from the file identified by the initially provided file name.
+func RemoteWatch(rf string, th int, v bool) (routing.DataClient, error) {
+	if !isFileRemote(rf) {
+		return Watch(rf), nil
+	}
+
+	tempFilename, err := ioutil.TempFile("", "routes")
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataClient := &remoteEskipFile{
+		remotePath: rf,
+		localPath:  tempFilename.Name(),
+		threshold:  th,
+		verbose:    v,
+	}
+
+	err = dataClient.DownloadRemoteFile()
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataClient.eskipFileClient = Watch(tempFilename.Name())
+
+	dataClient.preloaded = true
+
+	return dataClient, nil
+}
+
+// LoadAll returns the parsed route definitions found in the file.
+func (client *remoteEskipFile) LoadAll() ([]*eskip.Route, error) {
+
+	var err error = nil
+
+	if client.preloaded {
+		client.preloaded = false
+	} else {
+		err = client.DownloadRemoteFile()
+	}
+
+	if err != nil {
+		log.Errorf("LoadAll from remote %s failed. Continue using the last loaded routes", client.remotePath)
+		return nil, err
+	}
+
+	if client.verbose {
+		log.Debugf("New routes file %s was downloaded", client.remotePath)
+	}
+
+	return client.eskipFileClient.LoadAll()
+}
+
+// LoadUpdate returns differential updates when a remote file has changed.
+func (client *remoteEskipFile) LoadUpdate() ([]*eskip.Route, []string, error) {
+
+	err := client.DownloadRemoteFile()
+
+	if err != nil {
+		log.Errorf("LoadUpdate from remote %s failed. Trying to LoadAll", client.remotePath)
+		return nil, nil, err
+	}
+
+	newRoutes, deletedRoutes, err := client.eskipFileClient.LoadUpdate()
+	if err == nil {
+		if client.verbose {
+			log.Infof("New routes were loaded. New: %d; deleted: %d", len(newRoutes), len(deletedRoutes))
+
+			if client.threshold >= 0 {
+				if len(newRoutes) > client.threshold || len(deletedRoutes) > client.threshold {
+					log.Warnf("Significant amount of routes was updated. New: %d; deleted: %d", len(newRoutes), len(deletedRoutes))
+				}
+			}
+		}
+	} else {
+		log.Errorf("RemoteEskipFile LoadUpdate %s failed. Skipper continues to serve the last successfully updated routes. Error: %s",
+			client.remotePath, err)
+	}
+
+	return newRoutes, deletedRoutes, err
+}
+
+func isFileRemote(remotePath string) bool {
+	return strings.HasPrefix(remotePath, "http://") || strings.HasPrefix(remotePath, "https://")
+}
+
+func (client *remoteEskipFile) DownloadRemoteFile() error {
```

Comment:
i think the `DownloadRemoteFile` method is not used outside of this package, can we rename it to just lowercase `downloadRemoteFile`?

Reply:
We're using it from our `main.go`, so we "fail" (don't even do `skipper.run`) if the remote file is not downloadable: we prefer our cluster to keep "old" containers with a valid (even if maybe updated) route file in memory than to start with new but routeless instances.

Reply:
@xavivars we use the feature that Listener is only started after the first successful LoadAll() and health check from load balancer will keep us safe.
Additionally you can also test the endpoint with -status-checks

Reply:
Yes, the problem when I initially implemented this (more than one year ago) was that the feature you're talking wasn't working as we expected (due to the bug I reported & fixed few weeks ago). But agree, this shouldn't be needed anymore